aten/src/ATen/native/vulkan/ops/Tensor.h

```diff
@@ -0,0 +1,573 @@
+//    address space to maintain one single copy of the data, different access
+//    latencies from CPU and GPU to this shared location usually necessitates
+//    maintaining two copies each in processor-local memory, otherwise memory
+//    access latency will hurt from the processor to which this data is not
+//    close.  This shared memory is more often than not located in system memory,
+//    making for slow GPU read and write access over the PCI-e bus on discrete.
+//    Maintaining two separate copies on the other hand, requires synchronization
+//    to guarantee coherence.  This is not an issue on UMA and this implementation
+//    accounts for that optimization.
+//
+// 2) Second, synchronization across resources (i.e. buffers and images); GPU
+//    drivers pack images in proprietory formats for better locality of access
+//    and to enable lossless compression.  These conversions are both expensive
+//    (in general) and manual (in Vulkan.)  This requires a second order of
+//    synchronization to guarantee coherence between the contents of the buffer
+//    and image otherwise they will go out of sync.
+//
+// It is extremely important to keep in mind that the functionality this class
+// provides is generally expensive.  For optimal performance, the user of this
+// class should:
+//
+// 1) Avoid frequent CPU <=> GPU transfers which will be triggered if data is
+//    write accessed on one processor and read / write accessed on the other.
+//
+// 2) Avoid frequent buffer <=> image conversions which will be trigerred if
+//    data is write accessed as a buffer (image) and read accessed as an
+//    image (buffer).
+//
+// 3) When and if a synchronization is unavoidable, place as much distance
+//    between the synchronization is triggered and the data is accessed since
+//    all synchronizations this class provides are async.
+//
+// For optimal performance, access the data as images, and keep the data on GPU,
+// and above all understand the expensive data flow that this class abstracts
+// away.
+//
+// vTensor tries to address a specific concern and intentionally does not expose
+// GPU tensor memory directly.  Please keep that behavior intact as the whole
+// data model fundamentally depends on limiting what the user can achieve through
+// the interface to guarantee performance and coherence.
+//
+// A vTensor is associated with an api::Context as preparation for multi-GPU
+// support.
+//
+
+class vTensor final {
+ public:
+  vTensor() = default;
+  vTensor(
+      api::Context* context,
+      IntArrayRef sizes,
+      const TensorOptions& options);
+
+  /*
+    Types
+  */
+
+  typedef api::Resource::Memory::Access Access;
+  typedef api::Resource::Buffer Buffer;
+  typedef api::Resource::Fence Fence;
+  typedef api::Resource::Image Image;
+  typedef api::Resource::Memory Memory;
+
+  /*
+    Future
+  */
+
+  template<typename Type, Access::Flags kAccess>
```

Comment:
Nit: Usually I saw naming with k prefix for constexpr's. Am I missing something about the meaning of that k prefix? :)

Reply:
No, you are right.  I will revert.  :)